services/management/backup/artifacts_service.go

```diff
@@ -96,6 +100,104 @@ func (s *ArtifactsService) ListArtifacts(context.Context, *backupv1beta1.ListArt
 	}, nil
 }
 
+func (s *ArtifactsService) deleteArtifactRequirements(
+	artifactID string,
+) (string, *models.S3LocationConfig, error) {
+	var s3Config *models.S3LocationConfig
+	var artifactName string
+	if err := s.db.InTransaction(func(tx *reform.TX) error {
+		artifact, err := models.FindArtifactByID(tx.Querier, artifactID)
+		switch {
+		case err == nil:
+		case errors.Is(err, models.ErrNotFound):
+			return status.Errorf(codes.NotFound, "Artifact with ID %q not found.", artifactID)
+		default:
+			return err
+		}
+		if artifact.Status != models.SuccessBackupStatus && artifact.Status != models.ErrorBackupStatus {
+			return status.Errorf(codes.FailedPrecondition, "Artifact with ID %q isn't in the final state.", artifactID)
+		}
+
+		artifactName = artifact.Name
+
+		restoreItems, err := models.FindRestoreHistoryItems(tx.Querier, &models.RestoreHistoryItemFilters{
+			ArtifactID: artifactID,
+		})
+		if err != nil {
+			return err
+		}
+
+		for _, ri := range restoreItems {
+			if ri.Status == models.InProgressRestoreStatus {
+				return status.Errorf(codes.FailedPrecondition,
+					"Cannot delete artifact with ID %q: restore operation is in progress %q", artifactID, ri.ID)
+			}
+		}
+
+		location, err := models.FindBackupLocationByID(tx.Querier, artifact.LocationID)
+		if err != nil {
+			return err
+		}
+
+		s3Config = location.S3Config
+
+		return nil
+	}); err != nil {
+		return "", nil, err
+	}
+
+	return artifactName, s3Config, nil
+}
+
+// DeleteArtifact deletes specified artifact.
+func (s *ArtifactsService) DeleteArtifact(
+	ctx context.Context,
+	req *backupv1beta1.DeleteArtifactRequest,
+) (*backupv1beta1.DeleteArtifactResponse, error) {
+	artifactName, s3Config, err := s.deleteArtifactRequirements(req.ArtifactId)
+	if err != nil {
+		return nil, err
+	}
+
+	if req.RemoveFiles && s3Config != nil {
+		if err := s.s3.RemoveRecursive(
+			ctx,
+			s3Config.Endpoint,
+			s3Config.AccessKey,
+			s3Config.SecretKey,
+			s3Config.BucketName,
+			// Recursive listing finds all the objects with the specified prefix.
+			// There could be a problem e.g. when we have artifacts `backup-daily` and `backup-daily-1`, so
+			// listing by prefix `backup-daily` gives us both artifacts.
+			// To avoid such a situation we need to append a slash.
+			artifactName+"/",
+		); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := s.db.InTransaction(func(tx *reform.TX) error {
+		restoreItems, err := models.FindRestoreHistoryItems(tx.Querier, &models.RestoreHistoryItemFilters{
+			ArtifactID: req.ArtifactId,
+		})
+		if err != nil {
+			return err
+		}
+
+		for _, ri := range restoreItems {
+			if err := models.RemoveRestoreHistoryItem(tx.Querier, ri.ID); err != nil {
+				return err
+			}
+		}
+
+		return models.DeleteArtifact(tx.Querier, req.ArtifactId)
+	}); err != nil {
+		return nil, err
+	}
+
+	return &backupv1beta1.DeleteArtifactResponse{}, nil
```

Comment:
Whole this stuff should be done in one transaction.

Reply:
not sure it's a good idea to put the s3 call into the db transaction.

Reply:
reworked